tests/testutility.nim

```diff
@@ -71,3 +72,88 @@ suite "Utility":
   test "unsuccessful safeConvert from uint to int":
     check not (compiles do:
       result: uint = safeConvert[int, uint](11.uint))
+
+suite "withValue and valueOr templates":
+  type
+    TestObj = ref object
+      x: int
+
+  proc objIncAndOpt(self: TestObj): Opt[TestObj] =
+    self.x.inc()
+    return Opt.some(self)
+
+  proc objIncAndOption(self: TestObj): Option[TestObj] =
+    self.x.inc()
+    return some(self)
+
+  test "withValue calls right branch when Opt/Option is none":
+    var counter = 0
+    # check Opt/Option withValue with else
+    Opt.none(TestObj).withValue(v):
+      check false
+    else:
+      counter.inc()
```

Comment:
we can probably extract it to a proc inside the test and avoid the repetition

Reply:
You mean something like that?
```nim
proc testProc[T](val: Opt[T] | Option[T], counter: var int) =
  val.withValue(v):
    fail()
  else:
    counter.inc()
testProc(Opt.none(TestObj), counter)
testProc(none(TestObj), counter)
```
If there were more than two occurrences of this test I would do it, but I don't think it really brings something in this specific case.

Reply:
Yes, the benefit is avoiding code duplication, which I believe is never desirable.